src/back/hlsl/writer.rs

```diff
@@ -2675,13 +2733,175 @@ impl<'a, W: fmt::Write> super::Writer<'a, W> {
                         self.write_expr(module, arg, func_ctx)?;
                         write!(self.out, "))")?;
                     }
+                    // e: T,
+                    // newbits: T,
+                    // offset: u32,
+                    // count: u32
+                    // returns T
+                    // T is i32, u32, vecN<i32>, or vecN<u32>
+                    Function::InsertBits => {
+                        // mask = ((0xFFFFFFFFu >> (32 - count)) << offset)
+                        // return (count == 0 ? e : ((e & ~mask) | ((newbits << offset) & mask)))
+                        if let (Some(newbits), Some(offset), Some(count)) = (arg1, arg2, arg3) {
+                            write!(self.out, "(")?;
+                            self.write_expr(module, count, func_ctx)?;
+                            write!(self.out, " == 0 ? ")?;
+                            self.write_expr(module, arg, func_ctx)?;
+                            write!(self.out, " : ")?;
+                            write!(self.out, "(")?;
+                            self.write_expr(module, arg, func_ctx)?;
+                            write!(self.out, " & ~")?;
+                            // mask
+                            write!(self.out, "((0xFFFFFFFFu >> (32u - ")?;
+                            self.write_expr(module, count, func_ctx)?;
+                            write!(self.out, ")) << ")?;
+                            self.write_expr(module, offset, func_ctx)?;
+                            write!(self.out, ")")?;
+                            // end mask
+                            write!(self.out, ") | ((")?;
+                            self.write_expr(module, newbits, func_ctx)?;
+                            write!(self.out, " << ")?;
+                            self.write_expr(module, offset, func_ctx)?;
+                            write!(self.out, ") & ")?;
+                            // // mask
+                            write!(self.out, "((0xFFFFFFFFu >> (32u - ")?;
+                            self.write_expr(module, count, func_ctx)?;
+                            write!(self.out, ")) << ")?;
+                            self.write_expr(module, offset, func_ctx)?;
+                            write!(self.out, ")")?;
+                            // // end mask
+                            write!(self.out, "))")?;
+                        }
+                    }
+                    Function::Pack2x16float => {
+                        write!(self.out, "f32tof16(")?;
+                        self.write_expr(module, arg, func_ctx)?;
+                        write!(self.out, "[0]) | (f32tof16(")?;
+                        self.write_expr(module, arg, func_ctx)?;
+                        write!(self.out, "[1]) << 16)")?;
+                    }
+                    Function::PackBits {
```

Comment:
I was hoping to have less code duplication by putting the `Pack*` and `Unpack*` functions into PackBits/UnpackBits, but I didn't end up doing that as it seemed messier to have a bunch of `if`s scattered through the implementation vs just partially duplicating the code.

Should I move all the `Pack*` and `Unpack*` functions into their own implementations? (eg.                     `Function::Pack2x16unorm => {...}, Function::Pack4x8snorm => {...}, etc`)

Reply:
That would be preferred since it would eliminate the need for all the remaining `if`s. But do keep the dims/scale separate (don't inline them).

Reply:
Do you mean changing eg. [this block](https://github.com/Elabajaba/naga/blob/217b14bdbb9439ddd10e746a433bbfd9448f23be/src/back/hlsl/writer.rs#L2691) from

```rust
Mf::Pack2x16unorm => Function::PackBits {
    signed: false,
    dims: 2,
    scale: 65535,
},
```

to

```rust
Mf::Pack2x16unorm => Function::Pack2x16unorm {
    dims: 2,
    scale: 65535,
},
```